### PR TITLE
[MM-55270] fix(accessibility): "channels" button in global header is not selectable with keyboard

### PR DESCRIPTION
**Pull request by Victor-Nyagudi**
#### Summary

The product branding only handled click events to open the switcher menu. These changes make it possible to click on the product branding using <kbd>Space</kbd> and <kbd>Enter</kbd> to open/close the switcher menu.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/27068
Jira https://mattermost.atlassian.net/browse/MM-55270

#### Screenshots

<details><summary>Before</summary>

https://github.com/user-attachments/assets/518f5032-ce81-40b2-907f-1b484f1c54bd

</details> 

<details><summary>After</summary>

https://github.com/user-attachments/assets/3698f795-7d8a-4c49-979f-7bdc2745b35c

</details> 

#### QA Review Test Steps

1. Click the global header's empty portion to the search bar's left.
2. Press <kbd>Tab</kbd> until the product branding i.e. "Channels" is focused.
3. Press <kbd>Space</kbd> and verify it opens the switcher menu.
4. Press <kbd>Space</kbd> again and verify it closes the switcher menu.
5. Repeat steps 3 and 4 with the <kbd>Enter</kbd> key.
6. Repeat step 1 to remove focus and then left-click the product branding twice.
7. Verify this action toggles the switcher menu.

#### Release Note

```release-note
* Added the ability to toggle the switcher menu in the global header using the SPACE 
and ENTER keys while the product branding is in focus.
```

### CODE REVIEW

**Comment by Victor-Nyagudi** [comment]
Some useful information regarding this issue.

## Root Cause

If any other tag is used as a button, the `onclick` event only fires when the mouse clicks the tag. Keyboard events must be handled separately i.e. <kbd>Enter</kbd> and <kbd>Space</kbd>. The [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#:~:text=Buttons%20can%20be,key%20is%20pressed.) have more context on this.

An example of where a `div` clicked via the keyboard works is the search bar in the global header, which handles keyboard clicks separately. Interestingly, any button, not just the <kbd>Space</kbd>/<kbd>Enter</kbd> key, opens it.

## Proposed Solution

The [<kbd>Space</kbd> key event handler in `a11y_controller`](https://github.com/mattermost/mattermost/blob/ead2a7c0183fd786386c2c52a9e130d71366ed0a/webapp/channels/src/utils/a11y_controller.ts#L814) fires a click event whenever the space bar is pressed while a button is focused.

Another condition can be passed to the `if` statement so that whenever an element with the `role='button'` attribute experiences a <kbd>Space</kbd> key press event, it's clicked.

```diff
- if (event.target.nodeName === 'BUTTON') {
+ if (event.target.nodeName === 'BUTTON' || event.target.role === 'button') {
    event.preventDefault();
    event.stopPropagation();
    event.target.click();
}
```

[The same check](https://github.com/mattermost/mattermost/blob/ead2a7c0183fd786386c2c52a9e130d71366ed0a/webapp/channels/src/utils/a11y_controller.ts#L811) is done when the <kbd>Enter</kbd> key is pressed to click the element if it has the required role.

```JavaScript
case isKeyPressed(event, Constants.KeyCodes.ENTER):
    this.enterKeyIsPressed = true;
    if (event.target.role === 'button') {
    event.preventDefault();
    event.stopPropagation();
    event.target.click();
}
break;
```

The `div`s housing the product branding i.e. [`ProductBrandingTeamEdition`](https://github.com/mattermost/mattermost/blob/ead2a7c0183fd786386c2c52a9e130d71366ed0a/webapp/channels/src/components/global_header/left_controls/product_menu/product_branding_team_edition/product_branding_team_edition.tsx#L9) and [`ProductBranding`](https://github.com/mattermost/mattermost/blob/ead2a7c0183fd786386c2c52a9e130d71366ed0a/webapp/channels/src/components/global_header/left_controls/product_menu/product_branding/product_branding.tsx#L12) are then supplied with the `role='button'`. 

```JavaScript
// ...

const ProductBrandingTeamEditionContainer = styled.div.attrs(() => ({role: 'button'}))`

// ...
```

Adding this role also [tells screen readers the element is a button](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role#:~:text=Adding%20role%3D%22button%22%20tells%20the%20screen%20reader%20the%20element%20is%20a%20button%2C%20but%20provides%20no%20button%20functionality). Some useful information they didn't know about before therefore improving the accessibility even further.

## Other Solutions Explored

I turned the `div` of the `ProductBrandingTeamEdition` styled component into a button, which fixed the problem.

However, I had to make sure any elements inside the button resulted in valid HTML i.e. the `Badge` component is a `div`, so that had to be turned into a `span` for the HTML to be valid.

This approach only worked for the `ProductBrandingTeamEdition` component and not `ProdcutBranding` because that has an `h1` as a child element which would result in invalid HTML (an `h1` can't be inside a `button`).

I considered turning the `h1` into regular text and styling it to look like a heading and then having a legitimate `h1` styled out of view so screen readers can pick it up, but that felt a bit hacky and unreliable, so I scrapped the idea.

I also entertained the idea of passing an `onKeyDown` prop to `ProductBranding` and `ProductBrandingTeamEdition`. This prop is then used in a `useEffect` inside both components that handles a `keydown` event specifically for the <kbd>Enter</kbd> and <kbd>Space</kbd> keys.

When either of the keys is pressed, the function passed in `onKeyDown` is invoked to open the switcher menu. This only worked for the <kbd>Space</kbd> key. The <kbd>Enter</kbd> key rapidly opened and closed the switcher menu when used on the button next to the problematic logo, so this approach proved invalid.

My theory is the `onclick` handler in the `ProductMenuButton` also gets invoked at the same time, therefore the branding opens the switcher menu then the `ProductMenuButton` closes it right away.

Here's what I had in the `useEffect`. 

```JavaScript
// ...
 useEffect(() => {
        const handleKeyDown = (e: KeyboardEvent) => {
            if (Keyboard.isKeyPressed(e, Constants.KeyCodes.SPACE) || Keyboard.isKeyPressed(e, Constants.KeyCodes.ENTER)) {
                onKeyDown();
            }
        }

        document.addEventListener('keydown', handleKeyDown);

        return () => {
            document.removeEventListener('keydown', handleKeyDown);
        }
    }, []);
// ...
```

Finally, adding the `role='button'` to the `div` in the styled components of  `ProductBranding` and `ProductBrandingTeamEdition`  without any other changes resulted in <kbd>Space</kbd> opening and closing the switcher menu but not <kbd>Enter</kbd>. 

This most likely happens because the <kbd>Enter</kbd> key doesn't fire a click event in `a11y_controller`.

## Additional Notes

I'm curious why a ternary operator wasn't used in [this part of `ProductMenu`](https://github.com/mattermost/mattermost/blob/ead2a7c0183fd786386c2c52a9e130d71366ed0a/webapp/channels/src/components/global_header/left_controls/product_menu/product_menu.tsx#L118-L119).

```JavaScript
// ...
{license.IsLicensed === 'false' && <ProductBrandingTeamEdition/>}
{license.IsLicensed === 'true' && <ProductBranding/>}
// ...
```

A refactor might be necessary to tighten things up.

**Comment by Victor-Nyagudi** [comment]
Aside from two API tests that failed (I believe this is because of running the project locally) all the tests pass, so I'm not sure if any new e2e tests should be added. If so let me know.

I see an [`accessibility_keyboard_usability_spec.js`](https://github.com/mattermost/mattermost/blob/master/e2e-tests/cypress/tests/integration/channels/accessibility/accessibility_keyboard_usability_spec.js) file that looks like a good place to add the tests should they be needed.

**Comment by hmhealey** [comment]
Thanks for the PR, @Victor-Nyagudi, and thanks especially for the incredibly detailed description of the problem and the other solutions you tried as well.

My main concern with this approach is that we'll have two buttons in the heading that both open the product menu since they really should be a single button. We have another similar ticket (https://github.com/mattermost/mattermost/issues/26488) where we're trying to remove the need to tab through headings in the channel list twice, and this would cause that issue to happen here now. It's also a bit weird with this that only the button has hover and selected effects when that whole area of the header is clickable.

Because of that, I think putting the text into the button does actually end up being our best option. It would require an invisible h1 tag like you said, but I don't think that's actually as hacky of an idea as it seems. I've actually seen that recommended for cases somewhat like this one because you can also make the invisible h1 more useful by including things like the channel and team name, but that part is obviously outside the scope of this. You can use the `sr-only` class to make it invisible while keeping it visible to screen readers.

For the button itself, the `ProductMenuButton` is a button already, and it has a `label` prop that we can use to add that text. That prop is technically supposed to be a string, but looking at the code for it, it doesn't actually need to be a string, so we can pass the `ProductBranding`/`ProductBrandingTeamEdition` into it as long as we ignore the type check error using `// @ts-expect-error`. That part is definitely a hack, but I think it's also okay since we deprecated the Compass Components package which that comes from a while ago. You might have to play around with the CSS a bit to get the margin and text size to match the current version, but that should let you move everything into the button.

How does that sound?

@asaadmahmood Do you also have any thoughts on this since this would change how the product switcher menu works for everyone?

**Comment by Victor-Nyagudi** [comment]
> My main concern with this approach is that we'll have two buttons in the heading that both open the product menu since they really should be a single button.

This means having the product switcher icon and text inside one `<button>` i.e. merging the product branding text and product switcher button into one. It could work.

My thinking in the "Other Solutions Explored" involved having two separate buttons beside each other where the one with the text would require some changes, but I understand what you're saying.

> We have another similar ticket (https://github.com/mattermost/mattermost/issues/26488) where we're trying to remove the need to tab through headings in the channel list twice, and this would cause that issue to happen here now.

If the product switcher button and the product branding text are one button, there'll be no need to tab twice. 

You have to <kbd>Tab</kbd> twice to access both without my suggested changes, so the bug also occurs here.

> It's also a bit weird with this that only the button has hover and selected effects when that whole area of the header is clickable.

Once the two are merged, I'll need the new hover and selected states for the unified button. Someone from the design/UX team can be of assistance here.

> That prop is technically supposed to be a string, but looking at the code for it, it doesn't actually need to be a string, so we can pass the ProductBranding/ProductBrandingTeamEdition into it as long as we ignore the type check error using // @ts-expect-error. 

Do you mean pass the `ProductBranding`/`ProductBrandingTeamEdition` text to the `label` prop, or the whole component? I'm not sure how that would work.

Does all this sound right, @hmhealey?

**Comment by asaadmahmood** [comment]
@hmhealey I think I agree.
I don't really see a reason to have both toggle the same modal and also being different buttons.

@matthewbirtch may have more context on this.

**Comment by matthewbirtch** [comment]
> @hmhealey I think I agree. I don't really see a reason to have both toggle the same modal and also being different buttons.
> 
> @matthewbirtch may have more context on this.

Combining in to a single button makes sense, @asaadmahmood can you provide an updated hover state for this?

**Comment by hmhealey** [comment]
> Do you mean pass the ProductBranding/ProductBrandingTeamEdition text to the label prop, or the whole component? I'm not sure how that would work.

I mean the whole component. The `ProductMenuButton` uses an [`IconButton`](https://compass.mattermost.com/29be2c109/p/00b099-icon-button) which supports a `label` that it expects to be a string, but the code for it works fine if you supress the TS error like this.
```jsx
<ProductMenuContainer onClick={handleClick}>
    <ProductMenuButton
        active={switcherOpen}
        aria-expanded={switcherOpen}

        // @ts-expect-error
        label={<ProductBranding/>}
        aria-label={formatMessage({id: 'global_header.productSwitchMenu', defaultMessage: 'Product switch menu'})}
        aria-controls='product-switcher-menu'
    />
</ProductMenuContainer>
```

The text size and margin looks a bit off with that, but everything seems to work correctly.

@asaadmahmood For the hover state, the code above makes it use the same style as the Icon Button if that's what we'd want here

Screenshot 2024-11-20 at 3 24 30 PM

**Comment by Victor-Nyagudi** [comment]
> [!NOTE] 
> **TLDR:** The `ProductMenu` uses deprecated components, so if these will be migrated to the newer components, it's better to do that instead of joining the product switcher icon and text into one button here.  The recommended `Menu` handles the menu's open/close button differently but has its share of issues.

I've been chipping away at this for the past couple of days. Here's what I found.

The `ProductMenu` uses a deprecated [`MenuWrapper`](https://github.com/mattermost/mattermost/blob/dabf383d4cf38d8de985dfafca230ec8ae513967/webapp/channels/src/components/global_header/left_controls/product_menu/product_menu.tsx#L108) and [`Menu`](https://github.com/mattermost/mattermost/blob/dabf383d4cf38d8de985dfafca230ec8ae513967/webapp/channels/src/components/global_header/left_controls/product_menu/product_menu.tsx#L121) component. The warning recommends using the[ MUI `Menu`](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/components/menu/menu.tsx) instead.

Should this component be migrated to the Material UI menu, its button and the menu will be connected, so a better option would be to migrate it rather than combining the product switcher icon and text into one button now and then migrating later.

This comes with a fair amount of caveats. 

I found 35 components using the deprecated `MenuWrapper`. Each `Menu`'s child items must also be migrated to the newer `MenuItem` component. 

Rather than manually updating each component's menu item (there are plenty of them), I found it more efficient to update the `menuItem()` function to return the newer `MenuItem` since many components like  [`MenuItemLInk`](https://github.com/mattermost/mattermost/blob/dabf383d4cf38d8de985dfafca230ec8ae513967/webapp/channels/src/components/widgets/menu/menu_items/menu_item_link.tsx#L31) use this to return a deprecated menu item. 

https://github.com/mattermost/mattermost/blob/dabf383d4cf38d8de985dfafca230ec8ae513967/webapp/channels/src/components/widgets/menu/menu_items/menu_item.tsx#L12

All this worked well but required plenty of tweaking to get the styling to match what's currently in production since some styles are inherited, applied only if an element is a descendant of another, etc.

It wasn't ideal to make little changes in different places, but it wasn't messy (yet). However, if this is a precursor to what needs to be done to other components, a solid approach would make things easier.

My main concern during this process was MUI menu items don't work well with links or buttons inside them. The UP/DOWN navigation worked but pressing <kbd>Space</kbd> or <kbd>Enter</kbd> to follow a link or trigger a button did not.

Having links and buttons as direct descendants of a `ul` is also invalid HTML.

An [open issue](https://github.com/mui/material-ui/issues/33268) exists for this on the MUI repo. 

Overall, the lack of functionality of links and buttons inside MUI's `MenuItem` and the constant overriding of styles felt like a consistent struggle between Mattermost's native code and MUI. 

I think it would be best for the Mattermost team to create a strategy of either implementing the accessibility natively or relying solely on MUI to prevent the conflicts I encountered. What do you think?

If any of you would like to study my changes or tinker with them, here's the [branch](https://github.com/mattermost/mattermost/compare/master...Victor-Nyagudi:mattermost:fix-migrate-deprecated-menu) on my fork.

**Comment by hmhealey** [comment]
Sorry, I'm a bit lost here. Perhaps I'm misunderstanding, but it seems like we have two different issues here:
1. The "Channels" text isn't part of the menu button. That has an accessibility issue in that it seems like the "Channels" text should be focusable when it's not, and that could hopefully be mixed by adding it into that button
2. The `ProductMenu` using the deprecated Menu widget and not the newer MUI-based Menu component. The menu doesn't have proper keyboard support, but as far as I can tell, that's a separate issue. We also have https://mattermost.atlassian.net/browse/MM-55271 to work on that which we're actually planning on working on soon.

Is the deprecated Menu widget preventing us from moving the Channels text into the menu button and fixing that first issue? I wouldn't be incredibly surprised if it is, but I'm hoping we can address one issue at a time here.

Even if we don't migrate that menu yet right now, those notes will definitely be helpful when we get to migrating to the newer Menu component though

**Comment by Victor-Nyagudi** [comment]
@hmhealey My main concern was that since the `ProductMenuButton` and the `ProductMenu` are connected (one opens the other), I thought it would be a good idea to handle the button's accessibility and migrate to the newer menu. However, we can tackle one issue at a time. 

> Is the deprecated Menu widget preventing us from moving the Channels text into the menu button and fixing that first issue? 

The text can be moved to the menu button regardless of the deprecated `Menu`.

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for changing that. Ignoring the menu behaviour for now, I think it works great from the accessibility perspective.

It's a bit awkward that this issue falls in the middle of multiple different accessibility improvements and different pieces of deprecated code, so I appreciate both your willingness to tackle them all and to also walk some of that back so that we don't have changes that conflict with each other.

---

The only thing that I think is missing is the active/hover state of the `ProductMenuButton` which it got from the `IconButton` before. That's in [this section of the old component](https://github.com/mattermost/compass-components/blob/master/src/components/icon-button/IconButton.root.tsx#L82). There's a few things like the colours and `border-radius` which still need to be copied over.

I'll also reach out to double check that with Asaad since I think he might've missed the notification from GitHub

**Comment by asaadmahmood** [comment]
@Victor-Nyagudi I do not mind the hover state. That's fine, but I'd prefer a bit of padding/spacing changes for this button.
CleanShot 2024-12-12 at 21 44 24@2x

See here.
https://www.figma.com/design/iBH3jLcjap8eFebuKhG5AW/AI-Playbook-Statuses?node-id=3885-25939&node-type=frame&t=lNIiApsHWjiaeL36-11

**Comment by Victor-Nyagudi** [comment]
After making the necessary changes during different states, here's what the channel button looks like.


https://github.com/user-attachments/assets/5a086153-5b8b-42ee-b65b-b39b3b443179

**Review by hmhealey — CHANGES_REQUESTED** [review]
That looks good to me! Could you fix the indentation in that one file and the test snapshots?

I'll also run the E2E tests against the branch so that we can see any of those broke due to these changes

**Comment by Victor-Nyagudi** [comment]
@hmhealey Thanks for the reminder. All lint errors are fixed and I updated the snapshots.

By the way, adding a hover/active state revealed that the `FREE EDITION` badge is pushed down 1px leaving a slightly bigger space above it than below.

I only found out after recording this video (pause it at 2 seconds and observe).

https://github.com/user-attachments/assets/e58d5831-c407-4f9b-83f5-160ec4872366

<details><summary>Screenshot</summary>

mattermost-space-above-free-badge

</details> 

I [adjusted it](https://github.com/mattermost/mattermost/pull/29224/commits/44d9567677716b6a68887fa4d4f015aebd45288c), so everything's perfectly aligned now.

https://github.com/user-attachments/assets/36fb62cd-238c-4343-b061-9a038a7416ef

**Comment by hmhealey** [comment]
I tested the 2 test files that failed locally, and they're both passing there, so I think those are just flaky. I'll re-run them to see if that makes CI happy

/e2e-test

**Review by hmhealey — APPROVED** [review]

**Comment by marianunez** [comment]
@M-ZubairAhmed friendly reminder for this PR review

**Review by yasserfaraazkhan — APPROVED** [review]

**Review by M-ZubairAhmed — APPROVED** [review]
